test/src/test/java/jenkins/model/NodesTest.java

```diff
@@ -88,6 +89,17 @@ public void addNodeShouldReplaceExistingNode() throws Exception {
         assertThat(r.jenkins.getNode("foo"), sameInstance(newNode));
     }
 
+    @Test
+    @Issue("JENKINS-56403")
+    public void replaceNodeShouldRemoveExistingNode() throws Exception {
+        Node oldNode = r.createSlave("foo", "", null);
+        Node newNode = r.createSlave("foo-new", "", null);
+        r.jenkins.addNode(oldNode);
+        r.jenkins.getNodesObject().replaceNode(oldNode, newNode);
+        r.jenkins.getNodesObject().load();
+        assertNull(r.jenkins.getNode(oldNode.getNodeName()));
```

Comment:
I am not convinced this is really testing anything, since it is not using `RestartableJenkinsRule` to confirm that the behavior is correct _after_ a restart, which from my understanding was the reported issue. Whether calling `Nodes.load` programmatically like this actually has the same effect as restarting Jenkins is not obvious.

Reply:
@jglick The original issue was that the node was persisted on disk. It didn't specifically have anything to do with a restart, a "reload" would elicit the same behavior. So, this is a "tighter" functional test than one based on a restart .